ARTstor.js

```diff
@@ -1,131 +1,817 @@
+            }
+            //p=escape(p);
+            break;
+        case "collaboratoryfiltering":
+            var serviceURL = getServiceUrlRoot(url) + pageType + "/" + params[2] + "/thumbnails/" + startIdx + "/" + pageSize + "/" + sortOrder + "?collectionId=" + params[8];
+            break;
+        case "imagegroup": // image group
+        case "collections":
+        case "categories":
+        case "cluster":
+        default:
+            serviceURL = getServiceUrlRoot(url) + pageType + "/" + contentId + "/thumbnails/" + startIdx + "/" + pageSize + "/" + sortOrder;
+    }
+    return serviceURL;
+}
+
+function getServiceUrlRoot(url) {
+    var serviceRoot = getServerUrl(url) + "/secure/";
+    return serviceRoot;
+
+}
+
+function getFileRoot(url) {
+    var fileRoot = getServerUrl(url).substring(0, getServerUrl(url).lastIndexOf('/'));
+    return fileRoot;
+}
+
+/**
+    decodeSearchData is an helper function to process the search result page.
+    It converts the search url parameter into arrary of parameter values.
+
+    It converts
+    type=6&kw=stimson hall|all#and,1925|all&geoIds=&clsIds=&collTypes=&id=all&bDate=-2019&eDate=2030&dExact=1&prGeoId=&origKW=stimson hall|all#and,1925|all
+    to  {
+             "type": 6
+             "kw": "stimson hall|all#and,1925|all"
+             "geoIds": ""
+             "clsIds": ""
+             "collTypes": ""
+             "id": "all"
+             "bDate": "-2019"
+             "eDate": "2030"
+             "dExact": "1"
+             "prGeoId": ""
+             "origKW": "stimson hall|all#and,1925|all"
+         }
+**/
+function decodeSearchData(str) {
+    var searchParam = str.replace(/(&gt;)/g, ":");
+    var param = searchParam.split('&');
+    var searchData = new Object();
+    var sparam;
+    var id;
+    var value;
+    for (var i = 0; i < param.length; i++) {
+        sparam = param[i].split('=');
+        id = sparam[0];
+        value = sparam[1];
+
+        if (id === 'type') {
+            searchData[id] = parseInt(value);
+        } else {
+            searchData[id] = value;
+        }
+    }
+    return searchData;
+}
+
+var decodeMap = {
+    '20': ' ',
+    '21': '!',
+    '22': '"',
+    '23': '#',
+    '24': '$',
+    '25': '%',
+    '26': '&',
+    '27': "'",
+    '28': '(',
+    '29': ')',
+    '2A': '*',
+    '2B': '+',
+    '2C': ',',
+    '2D': '-',
+    '2E': '.',
+    '2F': '/',
+    '30': '0',
+    '31': '1',
+    '32': '2',
+    '33': '3',
+    '34': '4',
+    '35': '5',
+    '36': '6',
+    '37': '7',
+    '38': '8',
+    '39': '9',
+    '3A': ':',
+    '3B': ';',
+    '3C': '<',
+    '3D': '=',
+    '3E': '>',
+    '3F': '?',
+    '5B': '[',
+    //              '5C' : '\\',
+    '5D': ']',
+    '5E': '^',
+    '5F': '_',
+    '60': '`',
+    '7B': '{',
+    '7C': '|',
+    '7D': '}',
+    '7E': '~'
+};
+
+/**
+    Converts two character number character to special character
+    it converts 
+        type3D3626kw3Dairstream20trailer26geoIds3D
```

Comment:
there's something in your code about "!", which isn't explained here but looks like some sort of odd Unicode escapes in the format "!1234!" as decimal code point(?). I'm still not a fan of keeping mapping tables for something like hex character encodings. Here's a version of the decrypt function that should make this more maintainable (based on logic in your decrypt... I think)
```
function decrypt(s) {
  // Why not allow all hex values? Is the content ambiguously escaped? e.g. "C0"
  // If these kind of escapes are possible then everything falls apart anyway,
  // since something like "C325" would be escaped as "C25"
  return s.replace(/!(\d{1,5})!|(\d[\dA-F])/g, function(m, unicode, hex) {
    // Either unicode or hex are set, not both. unicode has priority over hex in the match
    if (unicode) return String.fromCharCode(parseInt(unicode)); // Always parses because of regexp

    // must be hex
    try {  return decodeURIComponent('%' + hex) }
    catch(e) { /* Some hex character escapes are invalid */ }

    return m; // Fail-safe
  });
}
```
This way you don't have to re-invent the wheel and worry about an incomplete mapping table.